plugins/modules/s3_object_info.py

```diff
@@ -0,0 +1,658 @@
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_attributes(connection, module, bucket_name, object_name)
+        elif key == 'object_legal_hold':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_legal_hold(connection, bucket_name, object_name)
+        elif key == 'object_lock_configuration':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_lock_configuration(connection, bucket_name)
+        elif key == 'object_retention':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_retention(connection, bucket_name, object_name)
+        elif key == 'object_tagging':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_tagging(connection, bucket_name, object_name)
+
+    return all_facts
+
+
+def get_object(connection, bucket_name, object_name):
+    params = {}
+    params['Bucket'] = bucket_name
+    params['Key'] = object_name
+
+    result = {}
+    object_info = {}
+
+    try:
+        object_info = connection.get_object(**params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        pass
+
+    if len(object_info) != 0:
+        # Remove Body, ResponseMetadata from object_info, convert to snake_case
+        del(object_info['Body'])
+        del(object_info['ResponseMetadata'])
+        object_info = camel_dict_to_snake_dict(object_info)
+
+    result['object_data'] = object_info
+
+    return result
+
+
+@AWSRetry.jittered_backoff(retries=10)
+def _list_bucket_objects(connection, **params):
+    paginator = connection.get_paginator('list_objects')
+    return paginator.paginate(**params).build_full_result()
+
+
+def list_bucket_objects(connection, module, bucket_name):
+    params = {}
+    params['Bucket'] = bucket_name
+
+    result = []
+    list_objects_response = {}
+
+    try:
+        list_objects_response = _list_bucket_objects(connection, **params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        pass
+
+    if len(list_objects_response) != 0:
+        # convert to snake_case
+        for response_list_item in list_objects_response['Contents']:
+            result.append(response_list_item['Key'])
+
+    module.exit_json(s3_keys=result)
+
+
+def main():
+
+    argument_spec = dict(
+        object_details=dict(type='dict', options=dict(
+            object_acl=dict(type='bool', default=False),
+            object_legal_hold=dict(type='bool', default=False),
+            object_lock_configuration=dict(type='bool', default=False),
+            object_retention=dict(type='bool', default=False),
+            object_tagging=dict(type='bool', default=False),
+            object_attributes=dict(type='bool', default=False),
+            attributes_list=dict(type='list', elements='str', choices=['ETag', 'Checksum', 'ObjectParts', 'StorageClass', 'ObjectSize']),
+        )),
```

Comment:
Fixed.

Reply:
I really think the best way to fix this is to use the arg spec as I suggested here, not to manually validate the provided params in the module. With the change you made, the module would go through the process of creating a client, checking if the bucket exists, and then checking if the object exists, only to then fail because of a parameter validation problem. This also means the error message the user gets will look different for this specific case of invalid parameters. Let ansible handle parameter validation. That's its job.

Reply:
Added.

Reply:
Thanks!